projector/binding.go

```diff
@@ -54,35 +58,95 @@ func New(mappingSource MappingSource) ServiceBindingProjector {
 }
 
 func (p *serviceBindingProjector) Project(ctx context.Context, binding *servicebindingv1beta1.ServiceBinding, workload runtime.Object) error {
-	mapping, err := p.mappingSource.LookupMapping(ctx, workload)
+	ctx, resourceMapping, version, err := p.lookupClusterMapping(ctx, workload)
 	if err != nil {
 		return err
 	}
-	mpt, err := NewMetaPodTemplate(ctx, workload, mapping)
+
+	// rather than attempt to merge an existing binding, unproject it
+	if err := p.Unproject(ctx, binding, workload); err != nil {
+		return err
+	}
+
+	versionMapping := MappingVersion(version, resourceMapping)
+	mpt, err := NewMetaPodTemplate(ctx, workload, versionMapping)
 	if err != nil {
 		return err
 	}
 	p.project(binding, mpt)
-	return mpt.WriteToWorkload(ctx)
+
+	if p.secretName(binding) != "" {
+		if err := p.stashLocalMapping(binding, mpt, resourceMapping); err != nil {
+			return err
+		}
+	}
+	if err := mpt.WriteToWorkload(ctx); err != nil {
+		return err
+	}
+
+	return nil
 }
 
 func (p *serviceBindingProjector) Unproject(ctx context.Context, binding *servicebindingv1beta1.ServiceBinding, workload runtime.Object) error {
-	mapping, err := p.mappingSource.LookupMapping(ctx, workload)
+	resourceMapping, err := p.retrieveLocalMapping(binding, workload)
+	if err != nil {
+		return err
+	}
+	ctx, m, version, err := p.lookupClusterMapping(ctx, workload)
 	if err != nil {
 		return err
 	}
-	mpt, err := NewMetaPodTemplate(ctx, workload, mapping)
+	if resourceMapping == nil {
+		// fall back to using the remote mappings, this isn't ideal as the mapping may have changed after the binding was originally projected
+		resourceMapping = m
+	}
+	versionMapping := MappingVersion(version, resourceMapping)
+	mpt, err := NewMetaPodTemplate(ctx, workload, versionMapping)
 	if err != nil {
 		return err
 	}
 	p.unproject(binding, mpt)
-	return mpt.WriteToWorkload(ctx)
+
+	if err := p.stashLocalMapping(binding, mpt, nil); err != nil {
+		return err
+	}
```

Comment:
This might be incorrect.  If this workload has multiple service bindings projecting into it, we can't remove the stashed mapping until every binding has unprojected itself.  Without synchronization on the workload, this is likely to race.

Reply:
Nevermind; on further review, the mapping is keyed off of the binding, so we're safe here.

Reply:
I went through that same thought process a couple times implementing it.